include/libp2p/protocol/kademlia/impl/asio_scheduler_impl.hpp

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef LIBP2P_ASIO_SCHEDULER_IMPL_HPP
+#define LIBP2P_ASIO_SCHEDULER_IMPL_HPP
+
+#include <boost/asio/deadline_timer.hpp>
+#include <boost/asio/io_context.hpp>
+#include <libp2p/protocol/kademlia/scheduler.hpp>
+
+namespace libp2p::protocol::kademlia {
+
+  class AsioSchedulerImpl : public Scheduler {
+   public:
+    AsioSchedulerImpl(boost::asio::io_context &io, Ticks interval)
+        : Scheduler{},
+          io_(io),
+          timer_(io, boost::posix_time::milliseconds(interval)),
+          interval_(interval),
+          started_(boost::posix_time::microsec_clock::local_time()),
+          timer_cb_([this](const boost::system::error_code &) { onTimer(); }),
+          immediate_cb_([this] { onImmediate(); }) {
```

Comment:
Do you really need those two callbacks?  If you want to invoke onTimer or onImmediate from timer you can use boost::bind for that. Example: https://github.com/soramitsu/kagome/pull/236/files#diff-e4975f9d1f0019308c8ea6f52a3c0e5eR115

Reply:
yes, those callbacks do completely different things